### PR TITLE
Use enum interface for GetRefs()

### PR DESCRIPTION
**Pull request by gerhardol**
https://github.com/gitextensions/gitextensions/pull/8990#discussion_r594314175

## Proposed changes

Replace the confusing GetRefs() interface with the enum interface introduced in #8990.
(A little too late change for 3.5).
Some optimizations for getting remotes, that had a special handling before.

## Test methodology 

Tests updated
Code review

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by mstv — COMMENTED** [review]
A very sensible refactoring.

Untouched code cannot be commented inline at GH - so here:

`GetRefsEnum` should be renamed to `GetRefsFlags`.

For the enum values, either stick to the (arguable) git naming (also used for `GitRefType`) or be descriptive:
`.Branches` means `.Heads` or `.LocalBranches`
`.Remotes` means `.RemoteBranches`

Typo in comment of `.All`: "abve".

File: GitUI/CommandsDialogs/FormDeleteBranch.cs [hunk 1]
```diff
@@ -42,7 +43,7 @@ public FormDeleteBranch(GitUICommands commands, IEnumerable<string> defaultBranc
 
         private void FormDeleteBranchLoad(object sender, EventArgs e)
         {
-            Branches.BranchesToSelect = Module.GetRefs(tags: true, branches: true).Where(h => h.IsHead && !h.IsRemote).ToList();
+            Branches.BranchesToSelect = Module.GetRefs(GetRefsEnum.Branches | GetRefsEnum.Tags).Where(h => h.IsHead).ToList();
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormDeleteBranch.cs:46]: Except for Remotes, I did not try to optimize the handling, just get something that worked as before.
I noticed this too, but concentrated on something that worked the same.
Will change.

File: GitUI/CommandsDialogs/FormMergeBranch.cs [hunk 2]
```diff
@@ -69,7 +70,7 @@ private void FormMergeBranchLoad(object sender, EventArgs e)
             var selectedHead = Module.GetSelectedBranch();
             currentBranchLabel.Text = selectedHead;
 
-            Branches.BranchesToSelect = Module.GetRefs(true, true);
+            Branches.BranchesToSelect = Module.GetRefs(GetRefsEnum.All);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormMergeBranch.cs:73]: If changed, Tags is not needed. There may be a few other such changes too (that was hard to understand with the previous API).
These changes should probably be separated to a separate commit.

File: GitUI/CommandsDialogs/FormPush.cs [hunk 3]
```diff
@@ -135,7 +135,7 @@ public FormPush(GitUICommands commands)
 
             void Init()
             {
-                _gitRefs = Module.GetRefs();
+                _gitRefs = Module.GetRefs(GetRefsEnum.All);
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormPush.cs:138]: same as above, resolving

File: GitUI/FilterBranchHelper.cs [hunk 4]
```diff
@@ -117,40 +118,12 @@ public void InitToolStripBranchFilter()
...
-            if (tag)
-            {
-                list.AddRange(GetTagsRefs());
-            }
+            GetRefsEnum refs = (local ? GetRefsEnum.Branches : GetRefsEnum.None)
+                | (tag ? GetRefsEnum.Tags : GetRefsEnum.None)
+                | (remote ? GetRefsEnum.Remotes : GetRefsEnum.None);
+            var list = Module.GetRefs(refs).Select(branch => branch.Name).ToList();
 
             return list;
```
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/FilterBranchHelper.cs:128]: ok

File: GitUI/CommandsDialogs/FormPush.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormPush.cs:138]: Only IsHead, IsRemote is used, changing

File: GitUI/CommandsDialogs/FormMergeBranch.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormMergeBranch.cs:73]: Similar in FormRebase (changing or commenting all GetRefsEnum.All occurrences).
So Tags should be included.
Rebasing on stash works for Git and can still be done in the revision grid (no need to block but also not a feature).
```suggestion
            // Offer rebase on refs also for tags (but not stash, notes etc)
            Branches.BranchesToSelect = Module.GetRefs(GetRefsEnum.Branches | GetRefsEnum.Remotes | GetRefsEnum.Tags);
```

**Comment by gerhardol** [comment]
> Untouched code cannot be commented inline at GH - so here:
> 
> `GetRefsEnum` should be renamed to `GetRefsFlags`.

GitRefsFlags

> 
> For the enum values, either stick to the (arguable) git naming (also used for `GitRefType`) or be descriptive:
> `.Branches` means `.Heads` or `.LocalBranches`
> `.Remotes` means `.RemoteBranches`

I would like to stick with the Git refs/ names so Heads, Remotes and Tags (and clarify at the definition)

> Typo in comment of `.All`: "abve".

OK

File: GitUI/CommandsDialogs/FormDeleteBranch.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormDeleteBranch.cs]: I may separate the Remotes changes to a separate commit too

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 5]
```diff
@@ -958,7 +958,8 @@ public void ForceRefreshRevisions()
 
                 _revisionReader ??= new RevisionReader();
 
-                var refs = Module.GetRefs(GetRefsEnum.All);
+                // Find all ambigous refs (including stash, notes etc)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:961]:
```suggestion
                // Find all ambiguous refs (including stash, notes etc)
```

**Review by mstv — APPROVED** [review]
👍

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs [hunk 6]
```diff
@@ -1,18 +1,21 @@
-﻿namespace GitCommands.Git.Commands
+﻿namespace GitUIPluginInterfaces
 {
     /// <summary>
     /// Enums requestable in GitRefs() (multiple names can be appended)
     /// Compare to <see ref="GitRefType"/> for actual values of parsed GitRefs
     /// </summary>
-    public enum GetRefsEnum
+    public enum GitRefsFlags
```
Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/GitRefsFlags.cs:7]: https://docs.microsoft.com/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces#naming-enumerations
```suggestion
    [Flags]
    public enum GitRefs
```

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
+        // Local Branches
+        Heads = 1 << 0,
+
+        // Remote branches
         Remotes = 1 << 1,
         Tags = 1 << 2,
 
-        // All refs, including those abve but also (at least) stash, notes and bisect
+        // All refs, including those above but also (at least) stash, notes and bisect
         All = 1 << 3
```
Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/GitRefsFlags.cs:19]:
```suggestion
        Everything = 1 << 3 | Heads | Remotes | Tags
```
...or rewrite the enum as
```cs
 Heads      = 0x0001,
 Remotes    = 0x0010,
 Tags       = 0x0100,
 Everything = 0x1111
```

Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/GitRefsFlags.cs:19]: Should we add "stash, notes and bisect" as individual entries as well?

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs:19]: Only add those we specifically need to get separately.

Any namespace can be added below refs/. 
Gerrit uses at least refs/for/ and refs/changes/ so all cannot be added.

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs (hunk 6 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs:7]: Can the guidelines be summarized somewhere of the current codestyle?
Even more important for new contributors.

Changing this, but the name is very close to the existing class GitRef, this is confusing.

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs (hunk 7 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: Prefer not to change this, then a separate flag is needed for All anyway.
The reason is that HasFlagButNotAll() need to be able to check if All bit is set, then the arguments for git-for-each-ref should be empty and that check cannot be done.

Sure, the most significant bit for All can be dynamically calculated in GetRefsCmd() too.
So the following (and another minor change) is required.
```c#
            static bool HasFlagButNotAll(GitRefs requested, GitRefs flag)
            {
                const GitRefs allOnly = GitRefs.All & ~(GitRefs.Heads | GitRefs.Remotes | GitRefs.Tags);
                return (requested & allOnly) == 0 && (requested & flag) != 0;
            }
```

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: I refactored GetRefsCmd() some to make this handling clearer.

File: GitCommands/Git/Commands/GitCommandHelpers.cs [hunk 8]
```diff
@@ -79,49 +79,29 @@ public static ArgumentString AddSubmoduleCmd(string remotePath, string localPath
...
                     : default)
             {
-                { sortBy != GitRefsSortBy.Default, GetSortCriteria(hasTags, sortBy, sortOrder), string.Empty },
-                format,
-                { HasFlag(getRef, GetRefsEnum.Branches), "refs/heads/", string.Empty },
-                { HasFlag(getRef, GetRefsEnum.Remotes), "refs/remotes/", string.Empty },
-                { HasFlag(getRef, GetRefsEnum.Tags), "refs/tags/", string.Empty },
+                sortCriteria(hasTags, sortBy, sortOrder),
+                gitRefsFormat(getRef, hasTags),
+                gitRefsPattern(getRef)
```
Comment (mstv) [inline GitCommands/Git/Commands/GitCommandHelpers.cs:93]:
```suggestion
                GetSortCriteria(hasTags, sortBy, sortOrder),
                GetGitRefsFormat(getRef, hasTags),
                GetGitRefsPattern(getRef)
```
?
Or else new convention for local functions?

**Review by mstv — APPROVED** [review]

File: GitCommands/Git/Commands/GitCommandHelpers.cs (hunk 8 above)
Comment (gerhardol) [reply GitCommands/Git/Commands/GitCommandHelpers.cs]: I had a squiggle to adjust the name that I acted on but do not see that now, changing back...

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs (hunk 6 above)
Comment (RussKie) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: > Can the guidelines be summarized somewhere of the current codestyle?

For many years in some shape or form I've been using Microsoft .NET design guidelines - originally [posted by Brad Abrams](https://docs.microsoft.com/en-us/archive/blogs/brada/internal-coding-guidelines) and the [current official](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/). 

A lot of styling guidelines can now be enforced with the help of StyleCop and Roslyn analyzers (and that what we've been doing for some time), though some things are hard to enforce automatically (like naming).

Comment (RussKie) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: > Changing this, but the name is very close to the existing class GitRef, this is confusing.

Perhaps we need to revisit the naming of the class or both. E.g. the enum can also be named `GitRefTypes`...

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs (hunk 7 above)
Comment (RussKie) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: 🤦‍♂️ Looks like I've made a clown of myself. Here's what I meant to write above:
```cs
[Flags]
public enum GitRefs : byte
{
	None       = 0,
	Heads      = 0x01,          // 00000001
	Remotes    = 0x02,          // 00000010
	Tags       = 0x04,          // 00000100
	Everything = byte.MaxValue  // 11111111
}
```


My point was that `All` makes no sense as a single bit. E.g. `GitRefs.Remotes | GitRefs.Tags` means "_remotes and tags_", but `GitRefs.All | GitRefs.Remotes | GitRefs.Tags` means "_everything_" with "remotes and tags" being totally superficial.

Since the members in the enum aren't representing the exhaustive set of values (which we may consider adding in the future), `All` (or `Everything`) must essentially be represented by "all bits set" value. And to assert whether it is set we just need to read the top bit, e.g. `((int)requested >> 7) == 1`

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: In this scenario, `Other` will not get any data at all or be the same as `Everything` which is confusing for an API usage.
The enums is the API, should be clear from a usage point of view.
This is also why I do not want to have a `None` value. I that value is used alone, the Git command should throw as it is meaningless, and how should `None|Tags` be interpreted?

All is a special mode comparable to Heads/Remotes/Tags even if All contains the latter. So `All|Tags` is the same as `All` regardless if all bits are set or not. Sure, that All is not used with other could be asserted, but that will be a runtime check without any real meaning.

I can add all bits to All but the internal check is a little weird in my opinion, the check must mask the enum to isolate the All bit.

Personally I find 1>>1, 1>>2 etc be easier to understand than listing hexvalues. (binary values work well here, harder when there are many values).

Comment (mstv) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: `GetRefs` uses this argument to control the filtering, where `All` means `Unfiltered`. Hence the enum value should be defined as `Unfiltered = 0`.

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: > `GetRefs` uses this argument to control the filtering, where `All` means `Unfiltered`. Hence the enum value should be defined as `Unfiltered = 0`.

Good suggestion, that should work
The name of the type should then be `GitRefsFilter`

Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: `Unfiltered|Tags` will be the same as `Tags` which could be confusing, that is API documentation.

Comment (mstv) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: > `Unfiltered|Tags` will be the same as `Tags` which could be confusing, that is API documentation.

Then `GitRefsFilter.None = 0`.

File: GitUI/FilterBranchHelper.cs [hunk 9]
```diff
@@ -117,42 +119,15 @@ public void InitToolStripBranchFilter()
...
-        }
-
         private List<string> GetBranchAndTagRefs(bool local, bool tag, bool remote)
         {
-            var list = GetBranchHeads(local, remote);
-            if (tag)
-            {
-                list.AddRange(GetTagsRefs());
-            }
+            // Options are interpreted as the refs the search should be limited too
```
Comment (gerhardol) [inline GitUI/FilterBranchHelper.cs:124]: This is a scope creep compared to master, I opted to add a comment instead of code to limit the functionality

**Comment by gerhardol** [comment]
Squashed without further changes.
I had planned to separate the optimizations to a separate commit, but after the last changes there will be more work than added value.

**Comment by gerhardol** [comment]
@msftbot squash in 3 days

**Review by mstv — APPROVED** [review]

File: Plugins/GitUIPluginInterfaces/GitRefsFlags.cs (hunk 7 above)
Comment (RussKie) [reply Plugins/GitUIPluginInterfaces/GitRefsFlags.cs]: Thank you @mstv - this makes sense :)

**Comment by RussKie** [comment]
MC

File: Plugins/GitUIPluginInterfaces/RefsFilter.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        // No filter, include those below but also (at least) stash, notes and bisect
+        // Note that if NoFilter is combined with a filter, the filter takes precedence
+        NoFilter = 0,
+
+        // Local Branches
+        Heads = 1 << 0,
+
+        // Remote branches
+        Remotes = 1 << 1,
+        Tags = 1 << 2
```
Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/RefsFilter.cs:21]:
```suggestion

        // Tags
        Tags = 1 << 2
```

**Review by RussKie — APPROVED** [review]

File: Plugins/GitUIPluginInterfaces/RefsFilter.cs (hunk 10 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/RefsFilter.cs:23]: Squashed this change (no other real merge conflict) when rebasing

**Comment by RussKie** [comment]
Merge at will
